llvm/labs/lab2/ivanov_nikita/ivanov.cpp

```diff
@@ -0,0 +1,66 @@
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+
+using namespace llvm;
+
+class MyLoopPass : public PassInfoMixin<MyLoopPass> {
+public:
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &FAM) {
+    auto *Ty = FunctionType::get(llvm::Type::getVoidTy(F.getContext()),
+                                 false); // get func type
+    auto *M = F.getParent();
+
+    auto LoopStartFunc =
+        M->getOrInsertFunction("loop_start", Ty); // create func start
+    auto LoopEndFunc =
+        M->getOrInsertFunction("loop_end", Ty); // create func end
+
+    auto &LI = FAM.getResult<LoopAnalysis>(F);
+
+    for (auto &L : LI) { // for by loop
+      auto *Preheader =
+          L->getLoopPreheader();           // get preheader block of the loop
+      auto *ExitBlock = L->getExitBlock(); // end exit block of the loop
+      IRBuilder<> Builder(Preheader->getTerminator()); // api for basic block
+
+      if (!Preheader)
```

Comment:
I'd suggest firstly perform all checks and if all is fine then insert `loop_*` calls.